### PR TITLE
fix(install): install binaries for packages installed multiple times

### LINKED ISSUE #1125: bun install should normalize line endings
### Version

All

### Platform

Linux 64bit

### What steps will reproduce the bug?

`bun install live-server`
`./node_modules/.bin/live-server`

Will raise this error:

> /usr/bin/env: ‘node\r’: No such file or directory

`npm install` normalizes the line endings

### How often does it reproduce? Is there a required condition?

_No response_

### What is the expected behavior?

_No response_

### What do you see instead?

> /usr/bin/env: ‘node\r’: No such file or directory

### Additional information

According to Jarred on discord:

> I looked at npm’s code and they do this specifically for bin files

### LINKED ISSUE #2516: Shebang on CRLF file accessed through a symlink throws error
### What version of Bun is running?

0.5.8

### What platform is your computer?

Linux 5.15.90.1-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

Install a package that defines a `bin` script, where the script file has CRLF line endings and begins with the `node` shebang:

```
#!/usr/bin/env node
```

Minimal reproduction: https://github.com/jakeboone02/bun-shebang-crlf-bug

1. `git clone https://github.com/jakeboone02/bun-shebang-crlf-bug.git`
2. `cd bun-shebang-crlf-bug`
3. `bun link` to register the package
4. `bun link bun-shebang-crlf-bug` to create symlinks in `./node_modules`
5. `bunx buncrlfbug` or `bunx --bun buncrlfbug`

### What is the expected behavior?

The script should run as expected, either with a Node instance when run with `bunx` or with Bun when run with `bunx --bun`.

### What do you see instead?

```
/usr/bin/env: ‘node\r’: No such file or directory
error: "buncrlfbug" exited with code 127
```

### Additional information

I first saw this issue when running

```sh
bunx jscodeshift ...
```

because `jscodeshift` ships with CRLF files. Converting all the `node_modules` files to LF avoids the problem (at least for the [example repo](https://github.com/jakeboone02/bun-shebang-crlf-bug)):

```sh
grep -lr $'\r$' node_modules | xargs -d '\n' sed -i 's/\r//g'
```

The CRLF issue does not seem to be a problem when symlinks are not involved.

### LINKED ISSUE #11799: Error: Bun failed to remap this bin to its proper location
### What version of Bun is running?

1.1.13+bd6a60512

### What platform is your computer?

Microsoft Windows NT 10.0.22631.0 x64

### What steps can reproduce the bug?

Dependencies:
```json
{
	"dependencies": {
		"@ariakit/react": "^0.4.7",
		"@clerk/remix": "4.1.0",
		"@clerk/themes": "2.1.9",
		"@conform-to/react": "1.1.4",
		"@conform-to/zod": "1.1.4",
		"@floating-ui/react": "0.26.16",
		"@floating-ui/react-dom": "2.1.0",
		"@fontsource-variable/nunito": "5.0.19",
		"@formkit/auto-animate": "0.8.2",
		"@radix-ui/react-scroll-area": "1.0.5",
		"@remix-run/node": "^2.9.2",
		"@remix-run/react": "^2.9.2",
		"@remix-run/serve": "^2.9.2",
		"@remix-run/server-runtime": "2.9.2",
		"@types/react": "npm:types-react@19.0.0-alpha.5",
		"@types/react-dom": "npm:types-react-dom@19.0.0-alpha.5",
		"@use-gesture/react": "10.3.1",
		"@vercel/remix": "2.9.2",
		"colorjs.io": "0.5.0",
		"convex": "1.12.1",
		"convex-ents": "0.7.7",
		"convex-helpers": "0.1.41",
		"date-fns": "3.6.0",
		"effect": "3.3.1",
		"hash-wasm": "4.11.0",
		"isbot": "^5.1.9",
		"iterator-helpers-polyfill": "3.0.1",
		"lodash-es": "4.17.21",
		"lucide-react": "0.394.0",
		"match-sorter": "^6.3.4",
		"node": "22",
		"random-word-slugs": "0.1.7",
		"react-textarea-autosize": "8.5.3",
		"react-twc": "1.4.1",
		"react-virtuoso": "4.7.11",
		"remix-routes": "1.7.5",
		"tailwind-merge": "2.3.0",
		"ts-dedent": "2.2.0",
		"zod": "3.23.8"
	},
	"devDependencies": {
		"@babel/preset-typescript": "7.24.7",
		"@biomejs/biome": "1.8.1",
		"@eslint/compat": "1.0.3",
		"@eslint/js": "9.4.0",
		"@mdx-js/rollup": "3.0.1",
		"@notionhq/client": "^2.2.15",
		"@remix-run/dev": "^2.9.2",
		"@tailwindcss/container-queries": "0.1.1",
		"@total-typescript/ts-reset": "0.5.1",
		"@types/bun": "1.1.3",
		"@types/eslint": "8.56.10",
		"@types/lodash-es": "4.17.12",
		"@types/node": "^20.14.2",
		"babel-plugin-react-compiler": "0.0.0-experimental-938cd9a-20240601",
		"dotenv-cli": "7.4.2",
		"eslint": "9.4.0",
		"eslint-plugin-react": "7.34.2",
		"eslint-plugin-react-compiler": "0.0.0-experimental-c8b3f72-20240517",
		"eslint-plugin-react-hooks": "4.6.2",
		"globals": "15.4.0",
		"npm-run-all": "4.1.5",
		"ora": "8.0.1",
		"prettier": "3.3.2",
		"prettier-plugin-jsdoc": "1.3.0",
		"prettier-plugin-organize-imports": "3.2.4",
		"prettier-plugin-tailwindcss": "0.6.3",
		"react": "19.0.0-rc-935180c7e0-20240524",
		"react-dom": "19.0.0-rc-935180c7e0-20240524",
		"rehype-autolink-headings": "7.1.0",
		"rehype-slug": "6.0.0",
		"remix-flat-routes": "0.6.5",
		"rollup-plugin-visualizer": "5.12.0",
		"tailwindcss": "3.4.4",
		"tailwindcss-animate": "^1.0.7",
		"ts-morph": "22.0.0",
		"tsx": "4.15.2",
		"typescript": "5.5.1-rc",
		"typescript-eslint": "8.0.0-alpha.29",
		"vite": "^5.2.13",
		"vite-plugin-babel": "1.2.0",
		"vite-plugin-inspect": "0.8.4"
	}
}
```

Run `bun install` and `bun install --force`

### What is the expected behavior?

Dependencies are installed without issues

### What do you see instead?

Received this output & error:
```
❯ bun install --force
[0.00ms] ".env.local"
bun install v1.1.13 (bd6a6051)
warn: incorrect peer dependency "eslint@9.4.0"

warn: incorrect peer dependency "eslint@9.4.0"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "@remix-run/dev@2.9.2"

warn: incorrect peer dependency "@remix-run/node@2.9.2"

warn: incorrect peer dependency "@remix-run/server-runtime@2.9.2"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react-dom@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"

warn: incorrect peer dependency "react@19.0.0-rc-935180c7e0-20240524"
  📦 Installing [947/1145]
error: failed to enqueue lifecycle scripts for esbuild: ENOENT
  📦 Installing [1066/1145] error: bin executable does not exist on disk

Bun failed to remap this bin to its proper location within node_modules.
This is an indication of a corrupted node_modules directory.

Please run 'bun install --force' in the project root and try
it again. If this message persists, please open an issue:
https://github.com/oven-sh/bun/issues


error: postinstall script from "esbuild" exited with 255
```

### Additional information

_No response_

### LINKED ISSUE #12113: Bun install regularly fails on windows
### How can we reproduce the crash?

bun install

### Relevant log output

bun install v1.1.13 (bd6a6051)
  📦 Installing [54/2402] ============================================================
Bun v1.1.13 (bd6a6051) Windows x64
Args: "C:\Users\**\.bun\bin\bun.exe" "i"
Features: bunfig dotenv extracted_packages(96)
Elapsed: 86698ms | User: 6296ms | Sys: 62546ms
RSS: 0.74GB | Peak: 0.86GB | Commit: 0.93GB | Faults: 528044

panic(main thread): Internal assertion failure
oh no: Bun has crashed. This indicates a bug in Bun, not your code.
### Stack Trace (bun.report)

Bun v1.1.13 ([`bd6a605`](<https://github.com/oven-sh/bun/tree/bd6a6051203ff577240615ff4e242270035a36b7>)) on windows x86_64 [InstallCommand]

**panic**: Internal assertion failure

- [`debug.zig:373`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/deps/zig/lib/std/debug.zig#L373>): `panicExtra`
- [`debug.zig:348`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/deps/zig/lib/std/debug.zig#L348>): `panic`
- [`output.zig:356`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/output.zig#L356>): `panic`
- [`bun.zig:3254`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/bun.zig#L3254>): `assertionFailure`
- [`types.zig:162`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/bun.js/node/types.zig#L162>): `unwrap`
- [`bin.zig:691`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/install/bin.zig#L691>): `link`
- [`install.zig:11372`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/install/install.zig#L11372>): `installPackageWithNameAndResolution`
- [`install.zig:12082`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/install/install.zig#L12082>): `installPackages`
- [`filter_run.zig:431`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/cli/filter_run.zig#L431>): `install`
- [`cli.zig:1465`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/src/cli.zig#L1465>): `start`

### PR DESCRIPTION
**Pull request by dylan-conway**
### What does this PR do?
Currently bun will link binaries for a package only once. This change will link binaries for each copy of the package.

Other changes in this pr:
- Links binaries in alphabetical order per `node_modules/.bin` folder for determinism.
- Skips linking binaries when the target does not exist
- removes usages of `getFdPath` in bin linking code. absolute paths are used (relative path is still used for link target)
- cleans up `.bunx` and `.exe` files when unlinking on windows
- normalizes shebang line ending

fixes #11799
fixes #12113 
fixes #1125 
fixes #2516 

### How did you verify your code works?
added a test

### CODE REVIEW

File: src/install/bin.zig [hunk 1]
```diff
@@ -266,429 +268,389 @@ pub const Bin = extern struct {
         }
     };
 
+    pub const PriorityQueueContext = struct {
+        lockfile: *const Install.Lockfile,
```
Comment (Jarred-Sumner) [inline src/install/bin.zig]:
```suggestion
        dependencies: *const std.ArrayList(Dependency),
```

**Comment by robobun** [comment]
:x: @dylan-conway, your commit 153c7344c16d156f349ad350a136a7bc025b6191 has 1 failures:

<li><code>buildkite-agent pipeline upload .buildkite/ci.yml</code> - code 1 on <a href="https://buildkite.com/bun/bun/builds/376#01905791-875d-474d-8a3a-d1a4cae67702">:gear:</a></li>

**Comment by robobun** [comment]
:x: @dylan-conway, your commit 153c7344c16d156f349ad350a136a7bc025b6191 has 1 failures:

<li><code>buildkite-agent pipeline upload .buildkite/ci.yml</code> - code 1 on <a href="https://buildkite.com/bun/bun/builds/376#01905791-875d-474d-8a3a-d1a4cae67702">:gear:</a></li>

File: src/resolver/resolve_path.zig [hunk 2]
```diff
@@ -299,7 +299,7 @@ pub fn longestCommonPathPosix(input: []const []const u8) []const u8 {
     return longestCommonPathGeneric(input, .posix);
 }
 
-threadlocal var relative_to_common_path_buf: [4096]u8 = undefined;
+threadlocal var relative_to_common_path_buf: bun.PathBuffer = undefined;
```
Comment (Jarred-Sumner) [inline src/resolver/resolve_path.zig:302]: This was intentionally larger than the max on macOS

**Comment by Jarred-Sumner** [comment]
There’s still one flaky test but merging

**Pull request merged**